File: Introductory Problems/palindrome reorder/palindrome.py
def palindrome_reorder(str): 
    alphaSet = []
    count = {}
    string = list(str.lower())
    for i in range(97, 123): 
        alphaSet.append(chr(i))
    for alpha in alphaSet: 
        if (alpha in string): 
            count[alpha] = string.count(alpha)
    if (len(string) % 2 != 0):
        oddSize = list(filter(lambda x: x % 2 != 0, count.values()))
        if (len(oddSize) != 1):
            return 'NO SOLUTION'
    else: 
        for freq in count.items(): 
            if (freq % 2 != 0): 
                return 'NO SOLUTION'
    return makePali(count)


def makePali(count): 
    start = "" 
    mid = ""
    end = ""
    for key in count: 
        if count[key] % 2 != 0:
            for _ in range(count[key]): 
                mid += key
        else: 
            for _ in range(int(count[key]/2)): 
                start += key
                end += key
    delim = ""
    end = list(end)
    end.reverse()
    end = delim.join(end)      
    return start+mid+end

print(palindrome_reorder('AAABACCBA'))

File: Introductory Problems/palindrome reorder/palindrome.js
function palindromeReorder(str) {
  const alphaSet = [],
    count = {};
  const string = Array.from(str.toLowerCase());
  for (let i = 97; i <= 122; i++) {
    alphaSet.push(String.fromCharCode(i));
  }
  for (let alpha of alphaSet) {
    if (string.includes(alpha)) {
      count[alpha] = string.reduce(
        (count, letter) => count + (letter == alpha ? 1 : 0),
        0
      );
    }
  }
  if (string.length % 2 != 0) {
    const oddSize = Object.values(count).filter((freq) => freq % 2 != 0).length;
    if (oddSize != 1) return "NO SOLUTION";
  } else {
    for (let frequency of Object.values(count)) {
      if (frequency % 2 != 0) return "NO SOLUTION";
    }
  }
  return makePali(count);
}

function makePali(count) {
  let start = "",
    end = "",
    mid = "";
  for (const key in count) {
    if (count[key] % 2 != 0) {
      mid = key.repeat(count[key]);
    } else {
      for (let i = 0; i < count[key] / 2; i++) {
        start += key;
        end += key;
      }
    }
  }
  end = Array.from(end).reverse().join("");
  return start + mid + end;
}

console.log(palindromeReorder("AAABACCBA"));
